File: erlang/bank-account/src/bank_account.erl
-module(bank_account).

-behaviour(gen_server).

-export([balance/1, charge/2, close/1, create/0, deposit/2, withdraw/2]).
-export([init/1, handle_call/3, handle_cast/2]).

-record(account, { state = open, balance = 0}).

%%%%%%%%%%%%%%%%%
%%% Interface %%%
%%%%%%%%%%%%%%%%%

create() ->
  {ok, Pid} = gen_server:start(?MODULE, [], []),
  Pid.

balance(Pid) ->
  gen_server:call(Pid, balance).

charge(Pid, Amount) ->
  gen_server:call(Pid, {charge, Amount}).

withdraw(Pid, Amount) ->
  gen_server:call(Pid, {withdraw, Amount}).

close(Pid) ->
  gen_server:call(Pid, close).

deposit(Pid, Amount) ->
  gen_server:cast(Pid, {deposit, Amount}).

%%%%%%%%%%%%%%%%%
%%% Callbacks %%%
%%%%%%%%%%%%%%%%%

init([]) ->
    {ok, #account{}}.

handle_call(_, _, #account{state = closed} = Account) ->
    {reply, {error, account_closed}, Account};

handle_call(balance, _, #account{balance = Balance} = Account) ->
    {reply, Balance, Account};

handle_call({charge, Amount}, _, #account{balance = Balance} = Account) ->
    if
        Amount < 0 -> {reply, 0, Account};
        Amount > Balance -> {reply, 0, Account};
        true -> {reply, Amount, Account#account{balance = Balance - Amount}}
    end;

handle_call({withdraw, Amount}, _, #account{balance = Balance} = Account) ->
    if
        Amount < 0 -> {reply, 0, Account};
        Amount > Balance -> {reply, Balance, Account#account{balance = 0}};
        true -> {reply, Amount, Account#account{balance = Balance - Amount}}
    end;

handle_call(close, _, Account) ->
    {reply, Account#account.balance, Account#account{state = closed}}.

handle_cast(_, #account{state = closed} = Account) ->
    {noreply, Account};

handle_cast({deposit, Amount}, #account{balance = Balance} = Account) when Amount > 0 ->
    {noreply, Account#account{balance = Balance + Amount}};

handle_cast({deposit, _Amount}, Account) ->
    {noreply, Account}.

File: java/bank-account/src/main/java/BankAccount.java
import java.util.concurrent.atomic.AtomicInteger;

class BankAccount {
    private AtomicInteger balance = null;

    void assertOpen(String message) throws BankAccountActionInvalidException {
        if (this.balance == null) {
            throw new BankAccountActionInvalidException(message);
        }
    }

    void assertPositive(int amount) throws BankAccountActionInvalidException {
        if (amount >= 0) return;
        throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount");
    }

    void open() throws BankAccountActionInvalidException {
        if (this.balance != null) {
            throw new BankAccountActionInvalidException("Account already open");
        }
        
        this.balance = new AtomicInteger();
    }

    void close() throws BankAccountActionInvalidException {
        assertOpen("Account not open");
        this.balance = null;
    }

    synchronized int getBalance() throws BankAccountActionInvalidException {
        assertOpen("Account closed");
        return this.balance.get();
    }

    synchronized void deposit(int amount) throws BankAccountActionInvalidException {
        assertOpen("Account closed");
        assertPositive(amount);
        this.balance.updateAndGet(balance -> amount + balance);
    }

    synchronized void withdraw(int amount) throws BankAccountActionInvalidException {
        assertOpen("Account closed");
        assertPositive(amount);

        int balance = this.balance.get();

        if (amount > balance) {
            throw new BankAccountActionInvalidException("Cannot withdraw more money than is currently in the account");
        }
        
        this.balance.set(balance - amount);
    }

}